api/screenshot/cache.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package screenshot
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/storage"
+
+	"github.com/web-platform-tests/wpt.fyi/shared"
+)
+
+func getHashesHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := shared.NewAppEngineContext(r)
+	ds := shared.NewAppEngineDatastore(ctx, false)
+
+	browser := r.FormValue("browser")
+	browserVersion := r.FormValue("browser_version")
+	os := r.FormValue("os")
+	osVersion := r.FormValue("os_version")
+
+	hashes, err := RecentScreenshotHashes(ds, browser, browserVersion, os, osVersion, nil)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	response, err := json.Marshal(hashes)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Write(response)
+}
+
+func uploadScreenshotHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "Only POST is supported", http.StatusMethodNotAllowed)
+		return
+	}
+
+	ctx := shared.NewAppEngineContext(r)
+	aeAPI := shared.NewAppEngineAPI(ctx)
+	// TODO(Hexcles): Abstract and mock the GCS utilities in shared.
+	gcs, err := storage.NewClient(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	bucketName := "wptd-screenshots-staging"
+	if aeAPI.GetHostname() == "wpt.fyi" {
+		bucketName = "wptd-screenshots"
+	}
+	bucket := gcs.Bucket(bucketName)
+
+	browser := r.FormValue("browser")
+	browserVersion := r.FormValue("browser_version")
+	os := r.FormValue("os")
+	osVersion := r.FormValue("os_version")
```

Comment:
nit: probably nicer to have a helper function that aggregates these into a product struct. (here and above)